File: mazes-for-programmers/python/src/mazes/recursive_backtracker.py
from random import choice
from grid import Grid
from image_saver import save
from colored_grid import ColoredGrid


class RecursiveBacktracker:
    @staticmethod
    def on(grid: Grid, start_at=None):
        while start_at is None:
            start_at = grid.random_cell()
        stack = [start_at]
        while stack:
            current = stack[-1]
            neighbors = [n for n in current.neighbors() if not n.links]
            if not neighbors:
                stack.pop()
            else:
                neighbor = choice(neighbors)
                current.link(neighbor)
                stack.append(neighbor)
        return grid


def recursive_backtracker_demo():
    grid = ColoredGrid(11, 11)
    RecursiveBacktracker.on(grid)
    save(grid.to_img(cell_size=20), "recursive_backtracker.png")

    middle = grid[grid.rows // 2, grid.columns // 2]
    grid.set_distances(middle.distances())

    save(grid.to_img(cell_size=20), "recursive_backtracker_colored.png")


# Demo
if __name__ == "__main__":
    recursive_backtracker_demo()

File: mazes-for-programmers/kotlin/src/main/kotlin/io/github/ocirne/mazes/RecursiveBacktracker.kt
package io.github.ocirne.mazes

class RecursiveBacktracker {

    companion object {
        fun on(grid: Grid, startAt: Cell = grid.randomCell()): Grid {
            val stack = ArrayDeque<Cell>()
            stack.addLast(startAt)
            while (stack.isNotEmpty()) {
                val current = stack.last()
                val neighbors = current.neighbors().filter { it!!.links.isEmpty() }

                if (neighbors.isEmpty()) {
                    stack.removeLast()
                } else {
                    val neighbor = neighbors.random()
                    current.link(neighbor!!)
                    stack.addLast(neighbor)
                }

            }
            return grid
        }
    }

    fun recursivelyOn(grid: Grid, start_at: Cell = grid.randomCell()): Grid {
        walkFrom(start_at)
        return grid
    }

    fun walkFrom(cell: Cell) {
        for (neighbor in cell.neighbors().shuffled()) {
            if (neighbor!!.links.isEmpty()) {
                cell.link(neighbor)
                walkFrom(neighbor)
            }
        }
    }
}
